common/src/main/java/org/apache/arrow/c/CometBufferImportTypeVisitor.java

```diff
@@ -0,0 +1,398 @@
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.List type) {
+    return Arrays.asList(maybeImportBitmap(type), importOffsets(type, ListVector.OFFSET_WIDTH));
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.LargeList type) {
+    return Arrays.asList(
+        maybeImportBitmap(type), importOffsets(type, LargeListVector.OFFSET_WIDTH));
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.FixedSizeList type) {
+    return Collections.singletonList(maybeImportBitmap(type));
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.Union type) {
+    switch (type.getMode()) {
+      case Sparse:
+        return Collections.singletonList(importFixedBytes(type, 0, UnionVector.TYPE_WIDTH));
+      case Dense:
+        return Arrays.asList(
+            importFixedBytes(type, 0, DenseUnionVector.TYPE_WIDTH),
+            importFixedBytes(type, 1, DenseUnionVector.OFFSET_WIDTH));
+      default:
+        throw new UnsupportedOperationException("Importing buffers for union type: " + type);
+    }
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.Map type) {
+    return Arrays.asList(maybeImportBitmap(type), importOffsets(type, MapVector.OFFSET_WIDTH));
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.Int type) {
+    return Arrays.asList(maybeImportBitmap(type), importFixedBits(type, 1, type.getBitWidth()));
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.FloatingPoint type) {
+    switch (type.getPrecision()) {
+      case HALF:
+        return Arrays.asList(
+            maybeImportBitmap(type), importFixedBytes(type, 1, /* bytesPerSlot= */ 2));
+      case SINGLE:
+        return Arrays.asList(
+            maybeImportBitmap(type), importFixedBytes(type, 1, Float4Vector.TYPE_WIDTH));
+      case DOUBLE:
+        return Arrays.asList(
+            maybeImportBitmap(type), importFixedBytes(type, 1, Float8Vector.TYPE_WIDTH));
+      default:
+        throw new UnsupportedOperationException("Importing buffers for type: " + type);
+    }
+  }
+
+  @Override
+  public List<ArrowBuf> visit(ArrowType.Utf8 type) {
+    try (ArrowBuf offsets = importOffsets(type, VarCharVector.OFFSET_WIDTH)) {
+      final int start = offsets.getInt(0);
+      final int end = offsets.getInt(fieldNode.getLength() * (long) VarCharVector.OFFSET_WIDTH);
+      checkState(
+          end >= start,
+          "Offset buffer for type %s is malformed: start: %s, end: %s",
+          type,
+          start,
+          end);
+      // HACK: For the issue https://github.com/apache/datafusion-comet/issues/540
+      // As Arrow Java doesn't support `offset` in C Data interface, we cannot correctly import
+      // a slice of string from arrow-rs to Java Arrow and then export it to arrow-rs again.
+      // So we add this hack to always take full length of data buffer by assuming the first offset
+      // is always 0 which is true for Arrow Java and arrow-rs.
+      final int len = end;
```

Comment:
This is the workaround added in Arrow Java side.

Reply:
A similar workaround is also added in the custom arrow-rs branch.

Reply:
This workaround assumes that the first element in offset buffer is always 0. This is true for arrow-rs and Arrow Java although Arrow spec just recommends it but not enforces it.

Reply:
Are the changes in the custom arrow-rs branch different to the fix that went into arrow-rs master branch recently?

Reply:
The fix into arrow-rs repo is real fix that correctly reflecting offset of offset buffer into `offset` field in C Data interface. But only the fix is not enough, we still need the `offset` support in Arrow Java.

So the workaround here is different but just hacky solution. Instead filing correct `offset` in C Data interface, we calculate correct data buffer length in a hacky way by assuming the first offset value in offset buffer is always 0. So we get correct data buffer length even offset buffer is a slice (i.e., `offset-ed` in arrow-rs).